python-client/giskard/llm/talk.py

```diff
@@ -0,0 +1,295 @@
+The value associated to the feature names found in `model_description` must be found calling any tool that are provided inside the 'Tools to gather information' list.
+If you fail to find a value after querying all the tool inside the 'Tools to gather information', you must return an explanation why you cannot answer with the list of missing feature that you need to answer'.
+
+Please note that you cannot share any confidential information with the user, and if the question is not related to the prediction model, you must return an explanation why you cannot answer.
+
+You can only use tool provided in the following list.
+"""
+MODEL_SUFFIX = """Begin!"
+
+Question: {input}
+Thought: I should look at the features that are required to see what I should gather to predict in the available tools
+{agent_scratchpad}"""
+
+
+class ModelSpec(BaseModel):
+    """Base class for model spec."""
+
+    model: GiskardBaseModel
+    dataset: Optional[Dataset]
+    scan_result: Optional[ScanResult]
+
+    def predict(self, text: str) -> str:
+        """Return the prediction of the model for a given row.
+
+        Args:
+            text: JSON representation of the row as a dict.
+        """
+        try:
+            features = json.loads(text)
+            # Filter only available features
+            features = {key: [features[key]] for key in features.keys() & self.model.meta.feature_names}
+            missing_features = {feature for feature in self.model.meta.feature_names if feature not in features.keys()}
+
+            if len(missing_features) > 0:
+                raise ValueError(f"Required feature `{missing_features}` have not been provided.")
+
+            return str(self.model.predict(Dataset(pd.DataFrame(features), validation=False)).prediction[0])
+        except Exception as e:
+            return repr(e)
+
+    def explain(self, text: str) -> str:
+        """Return an explanation of the prediction
+
+        Args:
+            text: JSON representation of the row as a dict.
+        """
+        if self.dataset is None:
+            return "Explanation is not available since no dataset has been provided"
+
+        try:
+            features = json.loads(text)
+            # Filter only available features
+            features = {key: features[key] for key in features.keys() & self.model.meta.feature_names}
+            missing_features = {feature for feature in self.model.meta.feature_names if feature not in features.keys()}
+
+            if len(missing_features) > 0:
+                raise ValueError(f"Required feature `{missing_features}` have not been provided.")
+
+            return str(explain(self.model, self.dataset, features))
+        except Exception as e:
+            return repr(e)
+
+    def scan_result_info(self):
+        if self.scan_result is None:
+            return "The model should be scanned with Giskard first"
```

Comment:
Can’t we just scan it now?

Reply:
I'm not sure if it's the right solution since it might take a lot of time to scan and the user will be waiting a lot of time to see the response.

Maybe we should add a confirmation message `I need to scan the model and it will take some time, do you want me to proceed?`. But the current implementation is not conversational but one shot